Change user updates to require explicit reactivation

senior opened the pull request:
Previously, attempting to update a user with the same email address as
an existing inactive user would trigger an implicit user reactivation
flow. This commit changes that implicit action to an explicit
one. Attempting to update an existing inactive user will now
fail. The `GET /` endpoint will not include all users, active and
inactive so that inactive users can be shown. This also creates a new
`/user/:id/reactivate` endpoint that will invoke the previously
implicit reactivation feature.

Based on what's in this PR and the designs from https://github.com/metabase/metabase/issues/4936#issuecomment-344109124 I think below is what still needs to be done.
- [x] Update the `People` page to include a Deactivated tab. Users from the `GET /api/users/` call will now include an `is_active` boolean and all users.
- [x] On the deactivated tab, add the reactivate link that uses the new `PUT /api/user/:id/reactivate` endpoint (along with the dialog)
- [x] The backend returns a 400 if the user attempts to update a user's email address to an existing address, but the frontend isn't propagating this error, that needs to be updated (see https://github.com/metabase/metabase/issues/7312)

iethree committed:
Change user updates to require explicit reactivation

Previously, attempting to update a user with the same email address as
an existing inactive user would trigger an implicit user reactivation
flow. This commit changes that implicit action to an explicit
one. Attempting to update an existing inactive user will now
fail. The `GET /` endpoint will not include all users, active and
inactive so that inactive users can be shown. This also creates a new
`/user/:id/reactivate` endpoint that will invoke the previously
implicit reactivation feature.

tlrobinson commented:
This is probably going to cause deactivated users to appear in the pulses recipient picker (and maybe other places?). Not sure if we should filter there or have a query parameter to `include_deactivated` or something.

tlrobinson commented:
@senior is there a way the frontend can get when a deactivated user was deactivated? That's in the designs.

tlrobinson commented on src/metabase/api/user.clj:
```diff
@@ -89,6 +83,17 @@
                                    is_superuser)))
   (User id))
 
+(api/defendpoint PUT "/:id/reactivate"
```
I think this should technically be a `POST` but I could be wrong

tlrobinson committed:
Implement deactivate/reactviate for people admin:

tlrobinson commented:
I've implemented the deactivated tab and reactivate action. Still need a way to see when the user was deactivated (include `updated_at`?) as well as surface errors when editing a user.

tlrobinson commented:
@senior Another problem is some errors in `PUT /api/user/:id` such as the `Email address already in use.` error don't follow our standard form error convention like `{ errors: { email: "Email address already in use." }}`

tlrobinson commented:
Also, `POST /api/user` doesn't return the `is_active` field so new users are appearing in the deactivated tab.

tlrobinson committed:
Surface errors when adding/editing users

tlrobinson commented on src/metabase/api/user.clj:
```diff
@@ -80,7 +73,8 @@
   ;; can't change email if it's already taken BY ANOTHER ACCOUNT
-  (api/check-400 (not (db/exists? User, :email email, :id [:not= id])))
+  (api/check (not (db/exists? User, :email email, :id [:not= id]))
+    [400 "Email address already associated to another user."])
```
This should conform to our errors conventions, either `{ errors: { email: "Email address ..." } }` if it's relevant to a specific field or `{ message: "Email address ..." }` if it applies to the whole request (the former in this case)

tlrobinson replied:
Copied from conversation with @camsaul:

> Writing all the stuff out is annoying so stick a helper function in `metabase.api.common` util like

```
(defn check-field [condition field-kw error-message]
  (when condition
    (throw (ex-info error-message {:status-code 400, :errors {field-kw error-message}}))))
```

> then you could do

```
(check-field (not (db/exists? User :email email)) :email (tru "Email address already in use."))
```

camsaul replied:
omg @tlrobinson I'm a moron I forgot we already have a function that does exactly that. It's called `api/checkp`. It takes the exact same params

tlrobinson commented on src/metabase/api/user.clj:
```diff
@@ -20,42 +20,35 @@
-                                   (session-api/google-auth-client-id)))
-      :ldap_auth     (boolean (and (:ldap_auth existing-user)
-                                   (ldap/ldap-configured?)))))
+  (db/select [User :id :first_name :last_name :email :is_superuser :google_auth :ldap_auth :last_login :is_active]))
```
How can we ensure `is_active` is returned from the `POST /api/user` and `PUT /api/user/:id` endpoints as well?

tlrobinson committed:
Add top level error message to EditUserForm

iethree committed:
Ensure is_active is included in user PUT/POST responses

iethree committed:
Add an `updated_at` column to `core_user`

iethree committed:
Switched user api errors to a map containing the error

senior commented:
@tlrobinson I added an `updated_at` field to `core_user`. `is_active` should now be included in the POST and PUT responses and the error messages should be fixed up as you suggested.

senior commented on src/metabase/api/user.clj:
```diff
@@ -19,43 +19,42 @@
-    ;; new user account, so create it
-    (user/invite-user! first_name last_name email password @api/*current-user*)))
-
+  (api/checkp (not (db/exists? User :email email))
```
@camsaul I wasn't aware of https://github.com/metabase/metabase/pull/7481#discussion_r186896077. Am I using the functions correctly here to get our standard error messages?

camsaul replied:
Yeah that looks right. I totally forgot this function existed.

camsaul replied:
It was previously more of an informal convention than anything else, however the `defendpoint` param validation code all returns errors in that format so we might as well make it consistent and try to do it everywhere

iethree committed:
Change GET `/api/user` to optionally return inactive users

iethree committed:
Merge remote-tracking branch 'upstream/master' into explicit-reactivation-of-users

senior commented:
Changed `GET /api/user` to return only active users by default. If `include_deactivated` is true, will include all users.

I also merge master in as there were a couple of conflicts. One of the conflicts was the migration I just added for `core_user.updated_at`. If you ran that migration before the merge you'll need to fix it up as it had an id of 76 before and now has an id of 77.

iethree committed:
Fix core_user.updated_at default value migration

tlrobinson committed:
Update people listing to show deactivated time, misc cleanup

tlrobinson committed:
Update snapshot

tlrobinson commented:
I've updated this with the things @senior added.

@senior could you merge `user.clj` again? Thanks.

@kdoh I'm not sure what's going on with this modal scrolling, I'm pretty sure it wasn't like that when I implemented it. Related to any of the recent flex changes?

![screenshot 2018-05-17 14 40 46](https://user-images.githubusercontent.com/18193/40206528-4bb19058-59e5-11e8-8286-27d2d2b96771.png)

tlrobinson committed:
Revert admin/people/people.js actions to thunk actions since they don't swallow errors

tlrobinson commented:
@senior it also looks like the backend tests are failing

kdoh commented:
@tlrobinson Yeah if I had to guess it's probably another nested `flex-full` that didn't actually need to be there. I can look into it if you'd like.

iethree committed:
Merge remote-tracking branch 'upstream/master' into explicit-reactivation-of-users

iethree committed:
Add underscore import to PeopleListingApp.jsx

iethree committed:
Remove unused namespace in metabase.middleware

kdoh committed:
remove nested flex-full

iethree committed:
Change the core_user.updated_at migration

MySQL 5.5 and before doesn't support computed default values for
datetime fields. It supports them in later versions, and on timestamp
fields for 5.5, but not datetime fields, which is what we use for our
auditing columns currently. This comment just removes that default
computed value and the related not null constraint.

tlrobinson committed:
Explicit reactivation: ensure is_active is always returned by /api/user endpoints

tlrobinson commented:
@senior It looks like `is_active` still wasn't being included in the `PUT /api/user`
PUT /api/user` response. I added it in my latest commit by just adding it to `default-user-fields` but that breaks a ton of backend test. If there's a better way could you add it to `PUT /api/user` (and revert my change)?

iethree committed:
Revert the addition of is_active in the default model props for user

senior commented:
@tlrobinson Could your backend be out of date? Just looked through the code and I have a test that covers this and it looks like it should all work. Maybe we can troubleshoot it together and figure out what's missing.

iethree committed:
Fix `/api/user` POSTs to return the same data as PUT

The previous code returned the result of the `db/insert!` this changes
that to user the same `fetch-user` function that is used elsewhere in
the user API to ensure we get the same data back (specifically looking
for `is_active`).

tlrobinson committed:
Merge branch 'master' of github.com:metabase/metabase into explicit-reactivation-of-users

tlrobinson committed:
Fix CI

tlrobinson committed:
Fix admin add user close button

tlrobinson commented:
Confirmed the last TODO is fixed. Also fixed a small bug with the cancel button not working. Once CI is green this should be ready.

tlrobinson committed:
Merge branch 'master' of github.com:metabase/metabase into explicit-reactivation-of-users

iethree committed:
Merge remote-tracking branch 'upstream/master' into explicit-reactivation-of-users
